### PR TITLE
Issues/835 forbid return values from test and lifecycle methods

### PR DESCRIPTION
**Pull request by mmerdes**
## Overview

This PR implements #835.

Following the existing styles in the respective parts of the codebase this is realised via predicates for `@Test` methods, but via `JUnitException`s for lifecycle methods. This might be subject to change in the context of the somewhat related issue #242.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+		assertTrue(contains(methods, "eight"));
+	}
+
+	private boolean contains(List<Method> methods, String methodName) {
+		// @formatter:off
+        return methods.stream()
+                .map(Method::getName)
+                .anyMatch(name -> name.equals(methodName));
+        // @formatter:on
+	}
```
Comment (jbduncan) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java]: @mmerdes I think the formatting is a bit off here. :)

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		assertEquals(2, methods.size());
+		assertTrue(contains(methods, "seven"));
+		assertTrue(contains(methods, "eight"));
+	}
+
+	private boolean contains(List<Method> methods, String methodName) {
+		// @formatter:off
+        return methods.stream()
+                .map(Method::getName)
+                .anyMatch(name -> name.equals(methodName));
```
Comment (jbduncan) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java]: @mmerdes I think it's possible to shorten this to `.anyMatch(methodName::equals)` if `methodName` is never null.

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java [hunk 3]
```diff
@@ -244,6 +244,11 @@ public static boolean isStatic(Member member) {
 		return Modifier.isStatic(member.getModifiers());
 	}
 
+	public static boolean returnsVoid(Method method) {
+		//checking the class itself yields problems with bridge methods
+		return method.getReturnType().toString().equals("void");
```
Comment (marcphilipp) [inline junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java]: `method.getReturnType().equals(void.class)`?

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java (hunk 2 above)
Comment (mmerdes) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java]: @jbduncan 
Thanks for the hint. Looks slightly better indeed.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java (hunk 1 above)
Comment (mmerdes) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/descriptor/LifecycleMethodUtilsTest.java]: @jbduncan Good call! It was a tabs vs spaces issue which got obscured by turning off the formatter...

File: junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java (hunk 3 above)
Comment (mmerdes) [inline junit-platform-commons/src/main/java/org/junit/platform/commons/util/ReflectionUtils.java]: Works like a charm. Opted for the `Void.TYPE` variant which I find slightly less ugly.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AfterAll.java [hunk 4]
```diff
@@ -36,6 +36,7 @@
  * {@link TestInstance @TestInstance(Lifecycle.PER_CLASS)}. {@code @AfterAll}
  * methods may optionally declare parameters to be resolved by
  * {@link org.junit.jupiter.api.extension.ParameterResolver ParameterResolvers}.
+ * {@code @AfterAll} methods must not return a value.
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/AfterAll.java]: I would move this new requirement to the first sentence and reword it something along the lines of:

```
{@code @AfterAll} methods must have a {@code void} return type, must not be {@code private}, and must be {@code static} by default.
```

And I would do the same for the rest of the related JavaDoc updates.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java [hunk 5]
```diff
@@ -46,6 +49,10 @@ public boolean test(Method candidate) {
 		if (isAbstract(candidate)) {
 			return false;
 		}
+		if (returnsVoid(candidate) != mustReturnVoid) {
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: I'd recommend:

```java
if (mustReturnVoid && !returnsVoid(candidate))
```

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/NonVoidTestableMethodIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+class NonVoidTestableMethodIntegrationTests {
+
+	@Test
+	void valid() {
+	}
+
+	//this method must never be called
+	@Test
+	int invalidMethodReturningPrimitive() {
+		assertTrue(false);
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/NonVoidTestableMethodIntegrationTests.java]: Instead of `assertTrue(false)`, I'd recommend invoking `fail()` with an appropriate failure message, and I'd do that for the rest of this test class.

Comment (mmerdes) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/NonVoidTestableMethodIntegrationTests.java]: Good call; saves the comments as well.

Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/NonVoidTestableMethodIntegrationTests.java]: exactly 😉

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java (hunk 5 above)
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: BTW, this requested change is not only about readability but more importantly about performance... since the suggested construct benefits from boolean short-circuiting.

Comment (mmerdes) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: True, that's more readable.

Comment (mmerdes) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: I guess the performance improvements would be hard to measure but the readability gain is substantial :)

Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: true! 😜

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AfterAll.java (hunk 4 above)
Comment (mmerdes) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/AfterAll.java]: ok

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java (hunk 5 above)
Comment (mmerdes) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: mmh, the more readable version creates a regression here:
`org.junit.jupiter.engine.discovery.predicates.IsTestFactoryMethodTests#factoryMethodReturningVoidEvaluatesToFalse`
which seems to be a valid test case to me...

Comment (mmerdes) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: In essence, the new behavior creates false-positives when we do have a void return type but need a collection or stream return type for factory methods. So will revert to the original behavior.

Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/predicates/IsTestableMethod.java]: hmmmm..... well, the "more readable version" seemed _logical_ to me, but if it's wrong... then it's wrong. 

Maybe I need to retake a basic course in boolean logic. 😉
